### PR TITLE
You can now cancel in the save prompt

### LINKED ISSUE #359: Be able to cancel in the save prompt on closing the editor

### PR DESCRIPTION
**Pull request by Sajdkick**
fixes #359

### CODE REVIEW

File: src/Engine/MainWindow.cpp [hunk 1]
```diff
@@ -91,7 +91,12 @@ bool MainWindow::ShouldClose() const {
 }
 
 void MainWindow::Close() {
-    shouldClose = true;
+	shouldClose = true;
+}
+
+void MainWindow::CancelClose() {
+	shouldClose = false;
```
Comment (NinjaDanz3r) [inline src/Engine/MainWindow.cpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Engine/MainWindow.cpp [hunk 2]
```diff
@@ -91,7 +91,12 @@ bool MainWindow::ShouldClose() const {
...
+	glfwSetWindowShouldClose(window, GLFW_FALSE);
```
Comment (NinjaDanz3r) [inline src/Engine/MainWindow.cpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Engine/MainWindow.cpp [hunk 3]
```diff
@@ -91,7 +91,12 @@ bool MainWindow::ShouldClose() const {
 }
 
 void MainWindow::Close() {
-    shouldClose = true;
+	shouldClose = true;
```
Comment (NinjaDanz3r) [inline src/Engine/MainWindow.cpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Engine/MainWindow.hpp [hunk 4]
```diff
@@ -70,6 +70,9 @@ class MainWindow {
         
         /// Close the window.
         void Close();
+
+		/// Cancel the closing of the window.
```
Comment (NinjaDanz3r) [inline src/Engine/MainWindow.hpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Engine/MainWindow.hpp [hunk 5]
```diff
@@ -70,6 +70,9 @@ class MainWindow {
...
+		void CancelClose();
```
Comment (NinjaDanz3r) [inline src/Engine/MainWindow.hpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/Editor.cpp [hunk 6]
```diff
@@ -90,9 +90,15 @@ void Editor::Show(float deltaTime) {
...
                 break;
 
-            case 1:
-                savePromptAnswered = true;
-                break;
+			case 1:
+				savePromptAnswered = true;
+				break;
+
+			case 2:
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.cpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/Editor.cpp [hunk 7]
```diff
@@ -90,9 +90,15 @@ void Editor::Show(float deltaTime) {
...
+				savePromptAnswered = false;
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.cpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/Editor.cpp [hunk 8]
```diff
@@ -90,9 +90,15 @@ void Editor::Show(float deltaTime) {
...
+				close = false;
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.cpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/Editor.cpp [hunk 9]
```diff
@@ -358,7 +364,11 @@ void Editor::Save() const {
 }
 
 bool Editor::ReadyToClose() const {
-    return savePromptAnswered;
+	return savePromptAnswered;
+}
+
+bool Editor::isClosing() const {
+	return close;
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.cpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/Editor.cpp [hunk 10]
```diff
@@ -90,9 +90,15 @@ void Editor::Show(float deltaTime) {
...
-                savePromptAnswered = true;
-                break;
+			case 1:
+				savePromptAnswered = true;
+				break;
+
+			case 2:
+				savePromptAnswered = false;
+				close = false;
+				savePromtWindow.ResetDecision();
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.cpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/Editor.cpp [hunk 11]
```diff
@@ -90,9 +90,15 @@ void Editor::Show(float deltaTime) {
...
+				break;
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.cpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/Editor.cpp [hunk 12]
```diff
@@ -90,9 +90,15 @@ void Editor::Show(float deltaTime) {
                 savePromptAnswered = true;
                 break;
 
-            case 1:
-                savePromptAnswered = true;
-                break;
+			case 1:
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.cpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/Editor.cpp [hunk 13]
```diff
@@ -90,9 +90,15 @@ void Editor::Show(float deltaTime) {
...
+				savePromptAnswered = true;
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.cpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/Editor.cpp [hunk 14]
```diff
@@ -90,9 +90,15 @@ void Editor::Show(float deltaTime) {
...
+				break;
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.cpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/Editor.cpp [hunk 15]
```diff
@@ -358,7 +364,11 @@ void Editor::Save() const {
 }
 
 bool Editor::ReadyToClose() const {
-    return savePromptAnswered;
+	return savePromptAnswered;
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.cpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/Editor.hpp [hunk 16]
```diff
@@ -34,11 +34,17 @@ class Editor {
         /// Loads the save editor state.
         void LoadEditorState();
 
-        /// Is the editor ready to be closed, have we saved everything?
-        /**
-        * @return Are we ready to close the editor?
-        */
-        bool ReadyToClose() const;
+		/// Is the editor ready to be closed, have we saved everything?
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.hpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/Editor.hpp [hunk 17]
```diff
@@ -34,11 +34,17 @@ class Editor {
...
+		/**
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.hpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/Editor.hpp [hunk 18]
```diff
@@ -34,11 +34,17 @@ class Editor {
...
+		 * @return Are we ready to close the editor?
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.hpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/Editor.hpp [hunk 19]
```diff
@@ -34,11 +34,17 @@ class Editor {
...
+		 */
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.hpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/Editor.hpp [hunk 20]
```diff
@@ -34,11 +34,17 @@ class Editor {
...
+		bool ReadyToClose() const;
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.hpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/Editor.hpp [hunk 21]
```diff
@@ -34,11 +34,17 @@ class Editor {
...
+
+		/// Is the editor preparing for closing down?
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.hpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/Editor.hpp [hunk 22]
```diff
@@ -34,11 +34,17 @@ class Editor {
...
+		/**
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.hpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/Editor.hpp [hunk 23]
```diff
@@ -34,11 +34,17 @@ class Editor {
...
+		 * @return Are we closing the editor?
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.hpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/Editor.hpp [hunk 24]
```diff
@@ -34,11 +34,17 @@ class Editor {
...
+		 */
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.hpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/Editor.hpp [hunk 25]
```diff
@@ -34,11 +34,17 @@ class Editor {
...
+		bool isClosing() const;
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.hpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/GUI/SavePromptWindow.cpp [hunk 26]
```diff
@@ -20,10 +20,11 @@ void SavePromptWindow::Show() {
             decision = 1;
             ImGui::CloseCurrentPopup();
         }
-        //if (ImGui::Button("cancel", ImVec2(120, 0))) {
-        //    decision = 2;
-        //    ImGui::CloseCurrentPopup();
-        //}
+		ImGui::SameLine();
```
Comment (NinjaDanz3r) [inline src/Editor/GUI/SavePromptWindow.cpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/GUI/ResourceView.cpp [hunk 27]
```diff
@@ -113,6 +113,12 @@ void ResourceView::Show() {
                     savePromptWindow.ResetDecision();
                     break;
 
+				case 2:
```
Comment (NinjaDanz3r) [inline src/Editor/GUI/ResourceView.cpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/GUI/ResourceView.cpp [hunk 28]
```diff
@@ -113,6 +113,12 @@ void ResourceView::Show() {
...
+					changeScene = false;
```
Comment (NinjaDanz3r) [inline src/Editor/GUI/ResourceView.cpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/GUI/ResourceView.cpp [hunk 29]
```diff
@@ -113,6 +113,12 @@ void ResourceView::Show() {
...
+					savePromptWindow.ResetDecision();
```
Comment (NinjaDanz3r) [inline src/Editor/GUI/ResourceView.cpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/GUI/ResourceView.cpp [hunk 30]
```diff
@@ -113,6 +113,12 @@ void ResourceView::Show() {
...
+					savePromptWindow.SetVisible(false);
```
Comment (NinjaDanz3r) [inline src/Editor/GUI/ResourceView.cpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/GUI/ResourceView.cpp [hunk 31]
```diff
@@ -113,6 +113,12 @@ void ResourceView::Show() {
...
+					break;
```
Comment (NinjaDanz3r) [inline src/Editor/GUI/ResourceView.cpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/main.cpp [hunk 32]
```diff
@@ -87,6 +87,9 @@ int main() {
     
                 editor->Show(deltaTime);
 
+				if (window->ShouldClose() && !editor->isClosing())
```
Comment (NinjaDanz3r) [inline src/Editor/main.cpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/main.cpp [hunk 33]
```diff
@@ -87,6 +87,9 @@ int main() {
...
+					window->CancelClose();
```
Comment (NinjaDanz3r) [inline src/Editor/main.cpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 36 issues
* ![MINOR][MINOR] 3 minor
* ![INFO][INFO] 33 info

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [main.cpp#L28](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/afe79ad3a7a9b2e737af404bc85e013341703628/src/Editor/main.cpp#L28): Return value of function freopen() is not used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AignoredReturnValue)
1. ![MINOR][MINOR] [main.cpp#L43](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/afe79ad3a7a9b2e737af404bc85e013341703628/src/Editor/main.cpp#L43): Variable 'vrSystem' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Editor/Editor.hpp [hunk 34]
```diff
@@ -35,11 +35,17 @@ class Editor {
         void LoadEditorState();
 
         /// Is the editor ready to be closed, have we saved everything?
-        /**
-        * @return Are we ready to close the editor?
-        */
+		/**
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.hpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

File: src/Editor/Editor.hpp [hunk 35]
```diff
@@ -35,11 +35,17 @@ class Editor {
...
+         * @return Are we ready to close the editor?
+         */
         bool ReadyToClose() const;
 
+        /// Is the editor preparing for closing down?
+		/**
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.hpp]: INFO horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 5 issues
* ![MINOR][MINOR] 3 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [main.cpp#L28](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/ab73e0da5391eaf968000166e642a5e64e55ea38/src/Editor/main.cpp#L28): Return value of function freopen() is not used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AignoredReturnValue)
1. ![MINOR][MINOR] [main.cpp#L43](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/ab73e0da5391eaf968000166e642a5e64e55ea38/src/Editor/main.cpp#L43): Variable 'vrSystem' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [main.cpp#L28](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/6dd24f8bac48779936ac71c1efa815503324a2fe/src/Editor/main.cpp#L28): Return value of function freopen() is not used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AignoredReturnValue)
1. ![MINOR][MINOR] [main.cpp#L43](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/6dd24f8bac48779936ac71c1efa815503324a2fe/src/Editor/main.cpp#L43): Variable 'vrSystem' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Engine/Component/Script.hpp [hunk 36]
```diff
@@ -31,7 +31,8 @@ namespace Component {
             /// The instance of the script class.
             asIScriptObject* instance = nullptr;
       
-                  /// Entities referenced by this script.
-                  std::vector<Entity*> refList;
+            /// Entities referenced by this script.
+            std::vector<Entity*> refList;
+
```
Comment (andens) [inline src/Engine/Component/Script.hpp]: Not necessary line break :)

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [main.cpp#L28](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/e1fec20d1ced1b556d5c2fceaeecf2b962f043b3/src/Editor/main.cpp#L28): Return value of function freopen() is not used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AignoredReturnValue)
1. ![MINOR][MINOR] [main.cpp#L43](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/e1fec20d1ced1b556d5c2fceaeecf2b962f043b3/src/Editor/main.cpp#L43): Variable 'vrSystem' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**